### DISCUSSION TITLE
Question on how the preprocessor is implemented

### DISCUSSION THREAD

**Discussion by shoedler** [microsoft/vscode-cpptools#8182] [Q&A] [answered]
I'm wondering how **cpptools** implements preprocessing of a single file (`#define`, `#include` etc. statements).

When hovering over a macro or a define occurence in a **.cpp* file, **cpptools** provides a `Hover` (vscode tooltip) containing the expansion of the macro / define:

![image](https://user-images.githubusercontent.com/38029550/134498540-8ee15501-1d25-4b99-902e-af737d34c8b0.png)

I just can't find out how this is implemented. I guess it is somewhere in the LSP, because the LSP is aware of the preprocessed file (no errors come up when typing something like in the image above)

I'd be happy to get some pointers on where to look in the codebase. 
Thanks in advance 😊

**Comment by sean-mcmanus** [comment, marked as answer]
All our parsing is done in our closed source cpptools process. Was there something specific you wanted to know about it?

**Reply by shoedler** [reply]
Thank you for the reply!

Well, I need to preprocess a file to resolve macros and the like, so that the parser of the language I'm working on receives correct data.

I was wondering which approach would be more efficient, maybe you could tell me how you did it in your cpptools process. 
Do you precompile using a external preprocessor, say **mcpp** or `cl.exe /P`? Or did you implement a precompiler within the cpptools process yourself?

**Reply by sean-mcmanus** [reply]
We use a parser based on https://www.edg.com/c (in some cases), but that would require you to purchase a license to access, so you might be better using an open source preprocessor, maybe from clang, or possibly calling the executable as you suggested.

**Reply by shoedler** [reply]
Thanks again Sean. I've opted to use **mcpp** in a child process and it looks promising
